src/libANGLE/renderer/d3d/d3d11/winrt/SwapChainNativeWindow.cpp

```diff
@@ -0,0 +1,182 @@
+            return false;
+        }
+    }
+
+    if (SUCCEEDED(result))
+    {
+        result = win.As(&mSwapChain);
+    }
+
+    if (SUCCEEDED(result))
+    {
+        // If a swapchain size is specfied, then the automatic resize
+        // behaviors implemented by the host should be disabled.  The swapchain
+        // will be still be scaled when being rendered to fit the bounds
+        // of the host.
+        // Scaling of the swapchain output is handled by the EGL client.
+        if (mSwapChainSizeSpecified)
+        {
+            mClientRect = {0, 0, swapChainSize.cx, swapChainSize.cy};
+        }
+        else
+        {
+            Size swapChainPanelSize;
+            result = GetSwapChainSize(mSwapChain, &swapChainPanelSize);
+
+            if (SUCCEEDED(result))
+            {
+                // Update the client rect to account for any swapchain scale factor
+                mClientRect = clientRect(swapChainPanelSize);
+            }
+        }
+    }
+
+    if (SUCCEEDED(result))
+    {
+        mNewClientRect     = mClientRect;
+        mClientRectChanged = false;
+        return true;
+    }
+
+    return false;
+}
+
+HRESULT SwapChainNativeWindow::createSwapChain(ID3D11Device *device,
+                                               IDXGIFactory2 *factory,
+                                               DXGI_FORMAT format,
+                                               unsigned int width,
+                                               unsigned int height,
+                                               bool containsAlpha,
+                                               IDXGISwapChain1 **swapChain)
+{
+    if (swapChain == nullptr || width == 0 || height == 0)
+    {
+        return E_INVALIDARG;
+    }
+    ASSERT(format == GetSwapChainFormat(mSwapChain));
+
+    HRESULT result = mSwapChain.CopyTo(swapChain);
+
+    // If the host is responsible for scaling the output of the swapchain, then
+    // scale it now before returning an instance to the caller.  This is done by
+    // first reading the current size of the swapchain panel, then scaling
+    if (SUCCEEDED(result))
+    {
+        if (mSwapChainSizeSpecified || mSwapChainScaleSpecified)
+        {
+            Size currentPanelSize = {};
+            result                = GetSwapChainSize(mSwapChain, &currentPanelSize);
+
+            // Scale the swapchain to fit inside the contents of the panel.
+            if (SUCCEEDED(result))
+            {
+                ASSERT(width == (unsigned int)currentPanelSize.Width);
+                ASSERT(height == (unsigned int)currentPanelSize.Height);
+                result = scaleSwapChain(currentPanelSize, mClientRect);
+            }
+        }
+    }
+
+    return result;
+}
+
+HRESULT SwapChainNativeWindow::scaleSwapChain(const Size &windowSize, const RECT &clientRect)
+{
+    return S_OK;
+}
+
+HRESULT GetSwapChainSize(const ComPtr<IDXGISwapChain> &swapChain, Size *windowSize)
```

Comment:
nit: should be in unnamed namespace

Reply:
why? I think if `GetSwapChainPanelSize()` in the `rx` namespace, so should this method be.

Reply:
> ...I don't see any reason for it. Anyway, there's no harm in keeping it here.

I don't either, and the small harm is having +1 symbol in a header instead of C++ file. So I decided to move into anon ns: f96225763efb5d65a725472546cfe91b7e234ea0.